### DISCUSSION TITLE
Best practice for simulating from priors under a given model

### DISCUSSION THREAD

**Discussion by fergusjchadwick** [inlabru-org/inlabru#189] [Ideas]
Hi All,

Thanks for all the work that goes into making inlabru! I'm trying to get to grips with inlabru coming from a Stan background and being fairly new to spatial stuff. One of the things that really helped me getting to grips with models (and especially specifying priors) in Stan (and JAGS before that!) was being able to do prior pushforward simulations (i.e. run the model without data). Is there a generic, best practice way to do this through inlabru or is it a case of hand coding in R? 

Best wishes,

Fergus

**Comment by finnlindgren** [comment]
Hi Fergus!

There is unfortunately no fully automated way to do this; it's been on the wishlist for INLA/inlabru for a long time though.  INLA does have most of the information about the parameter priors (`INLA::inla.priors.used()`) and latent model precisions avaliable (e.g. `INLA::inla.spde2.precision()` for spde models, and sampling can be done with `INLA::inla.qsample()`), so in principle it can be done, but we do need a proper interface for automating it.

Part of the difficulty is that some aspects of the model are automatically determined only when inla() is called for estimation, so one may need to do a "dry run" of inla first (an experimental `inla.dryrun()` function was recently added for related reasons).
Ideally, there should be a `prior=TRUE` option for `inlabru::generate()`, and a corresponding `INLA::inla.prior.sample()` function.

Another complication is that many practical spatial models involve intrinsic stationary random fields, that without constraining them would have infinite variance; to sample, one either have to constrain them in some way, or not use such models.  However, this issue also affects "fixed effect" covariate components, that a priori have a very large variance.  In practice, one might want to only sample individual model components, that have well defined distributions a priori.


Finn

**Reply by fergusjchadwick** [reply]
Hey Finn,

Thanks for the quick response! 

The interface you mention would be ideal (I'm happy to be a beta tester for new functionality if that would be helpful!) but it sounds like there are good alternatives to try in the meanwhile.

I hadn't thought about the latter - can the constraints be imposed in a meaningful way? i.e. is that actually a useful behaviour to expose in the prior?

Fergus

**Reply by finnlindgren** [reply]
One can sample with constraints; that will at least provide information about typical spatial patterns of those models.  One just needs to be mindful of the differences between constrained and unconstrained models.

**Reply by fergusjchadwick** [reply]
That makes sense. Is it also possible to use a more constrained prior for these things? Apologies if this is a basic question!

**Reply by finnlindgren** [reply]
The intrinsic stationary field models are intrinsically ill-determined, unless you impose a constraint. But for variance parameters etc, then yes, using more constraining priors is usually what one ends up doing if one wants prior samples to be reasonable.
